src/Waives.Pipelines/Pipeline.cs

```diff
@@ -226,26 +211,68 @@ public Pipeline OnDocumentError(Action<DocumentError> action)
         /// this object.</returns>
         public IDisposable Start()
         {
-            _pipeline = _pipeline.SelectMany(async d =>
+            var (disposable, _) = StartAsync();
+            return disposable;
+        }
+
+        private (IDisposable, Task) StartAsync()
+        {
+            // take a copy of _docActions so Start() is idempotent
+            var docActions = new List<Func<WaivesDocument, Task<WaivesDocument>>>(_docActions)
             {
-                await d.HttpDocument.Delete(() =>
+                async d =>
                 {
-                    _rateLimiter.MakeDocumentSlotAvailable();
-                    _logger.Log(LogLevel.Info, $"Completed processing '{d.Source.SourceId}' and deleted Waives document");
-                }).ConfigureAwait(false);
+                    await d.HttpDocument.Delete().ConfigureAwait(false);
 
-                return d;
-            });
+                    _logger.Log(LogLevel.Info,
+                        $"Completed processing '{d.Source.SourceId}' and deleted Waives document");
+
+                    return d;
+                }
+            };
+
+            var taskCompletion = new TaskCompletionSource<bool>(
+                TaskCreationOptions.RunContinuationsAsynchronously);
+
+            void OnPipelineComplete()
+            {
+                _onPipelineCompletedUserAction();
+                _logger.Log(LogLevel.Info, "Pipeline complete");
+                taskCompletion.SetResult(true);
+            }
+
+            void OnDocumentException(Exception exception, Document document)
+            {
+                _onDocumentError(new DocumentError(document, exception));
```

Comment:
💪 Should we set `taskCompletion`'s result to `false` in here if `_onDocumentError` throws?

Reply:
Any PipelineExceptions are now propagated back to the pipeline from the observer,  and `taskCompletion.SetException()` is called.

Reply:
Nice one :+1: